File: Java/Motor Follower/src/main/java/frc/robot/Robot.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package frc.robot;

import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.TimedRobot;

import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.config.SparkMaxConfig;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.SparkBase.ResetMode;


/**
 * This sample program shows how to control a motor using a joystick. In the
 * operator control part of the program, the joystick is read and the value is
 * written to the motor.
 *
 * <p>Joystick analog values range from -1 to 1 and speed controller inputs also
 * range from -1 to 1 making it easy to work together.
 */
public class Robot extends TimedRobot {
  private static final int leadDeviceID = 1;
  private static final int followDeviceID = 2;
  private static final int kJoystickPort = 0;

  private SparkMax leadMotor;
  private SparkMax followMotor;
  private SparkMaxConfig followMotorConfig;
  private Joystick joystick;

  @Override
  public void robotInit() {
    /**
     * SPARK MAX controllers are intialized over CAN by constructing a CANSparkMax object
     * 
     * The CAN ID, which can be configured using the SPARK MAX Client, is passed as the
     * first parameter
     * 
     * The motor type is passed as the second parameter. Motor type can either be:
     *  com.revrobotics.spark.SparkLowLevel.MotorType.kBrushless
     *  com.revrobotics.spark.SparkLowLevel.MotorType.kBrushed
     * 
     * The example below initializes two brushless motors with CAN IDs 1 and 2. Change
     * these parameters to match your setup
     */
    leadMotor = new SparkMax(leadDeviceID, MotorType.kBrushless);
    followMotor = new SparkMax(followDeviceID, MotorType.kBrushless);
    /**
     * We modify the followMotor's SPARK MAX configurator to follow the leadMotor by
     * calling the follow() method and passing both the leader CAN ID and the invert value.
     * If only the leader CAN ID is passed, invert will default to false. 
     * 
     * NOTE: Changes made in the configurator do not get applied until the configurator is passed 
     * to the configure() method.
     */
    followMotorConfig = new SparkMaxConfig();
    followMotorConfig.follow(leadDeviceID, false);

    /**
     * The configure() method will apply the parameter changes to the motor its called from.
     * 
     * The first parameter is the configurator object which will contain all the changes
     * we want to apply.
     * 
     * The second parameter is the reset mode and can be either:
     *    kResetSafeParameters - Restore defaults before applying changes
     *    kNoResetSafeParameters - Don't restore defaults before applying changes
     * 
     * The third parameter is the persist mode and can be either:
     *    kPersistParameters -  Save settings onto SPARK MAX EEPROM
     *    kNoPersistParameters -  Don't save settings onto SPARK MAX EEPROM
     * 
     */
    followMotor.configure(followMotorConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

    joystick = new Joystick(kJoystickPort);
  }

  @Override
  public void teleopPeriodic() {
    /**
     * followMotor will automatically follow whatever the applied output is on leadMotor.
     * 
     * Thus, set only needs to be called on leadMotor to control both of them
     */
    leadMotor.set(joystick.getY());
  }
}

File: C++/Arcade Drive With CAN/src/main/cpp/Robot.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <frc/Joystick.h>
#include <frc/TimedRobot.h>
#include <frc/drive/DifferentialDrive.h>
#include "rev/SparkMax.h"
#include "rev/config/SparkMaxConfig.h"

class Robot : public frc::TimedRobot {
  /**
   * SPARK MAX controllers are initialized over CAN by constructing a SparkMax object
   * 
   * The CAN ID, which can be configured using the SPARK MAX Client, is passed as the
   * first parameter
   * 
   * The motor type is passed as the second parameter. Motor type can either be:
   *  kMotorType
   *  rev::spark::SparkMax::MotorType::kBrushed
   * 
   * The example below initializes four brushless motors with CAN IDs 1, 2, 3 and 4. Change
   * these parameters to match your setup
   */
  static constexpr int kLeftLeadDeviceID = 1, kLeftFollowDeviceID = 2, kRightLeadDeviceID = 3, kRightFollowDeviceID = 4;
  static constexpr auto kMotorType = rev::spark::SparkMax::MotorType::kBrushless;
  rev::spark::SparkMax m_leftLeadMotor{kLeftLeadDeviceID, kMotorType};
  rev::spark::SparkMax m_rightLeadMotor{kRightLeadDeviceID, kMotorType};
  rev::spark::SparkMax m_leftFollowMotor{kLeftFollowDeviceID, kMotorType};
  rev::spark::SparkMax m_rightFollowMotor{kRightFollowDeviceID, kMotorType};

  /**
   * In RobotInit() below, we will configure m_leftFollowMotor and m_rightFollowMotor to follow
   * m_leftLeadMotor and m_rightLeadMotor, respectively.
   *
   * Because of this, we only need to pass the lead motors to m_robotDrive. Whatever commands are
   * sent to them will automatically be copied by the follower motors
   */
  frc::DifferentialDrive m_robotDrive{m_leftLeadMotor, m_rightLeadMotor};

  frc::Joystick m_stick{0};
  
 public:
  void RobotInit() {
    /**
     * The ResetMode::kResetSafeParameters constant can be used to reset
     * the configuration parameters in the SPARK MAX to their factory
     * default state. If PersistMode::kNoPersistParameters is passed,
     * these parameters will not persist between power cycles.
     */
    rev::spark::SparkMaxConfig leftLeadConfig;

    m_leftLeadMotor.Configure(leftLeadConfig,
      rev::spark::SparkMax::ResetMode::kResetSafeParameters,
      rev::spark::SparkMax::PersistMode::kNoPersistParameters);

    rev::spark::SparkMaxConfig rightLeadConfig;

    m_rightLeadMotor.Configure(rightLeadConfig,
      rev::spark::SparkMax::ResetMode::kResetSafeParameters,
      rev::spark::SparkMax::PersistMode::kNoPersistParameters);

    /**
     * In CAN mode, one SPARK MAX can be configured to follow another. This is done by calling
     * the Follow() method on the SPARK MAX Config you want to configure as a follower, and by passing
     * as a parameter the SPARK MAX you want to configure as a leader.
     * 
     * This is shown in the example below, where one motor on each side of our drive train is
     * configured to follow a lead motor.
     */
    rev::spark::SparkMaxConfig leftFollowConfig;
    leftFollowConfig.Follow(m_leftLeadMotor);

    m_leftFollowMotor.Configure(leftFollowConfig,
      rev::spark::SparkMax::ResetMode::kResetSafeParameters,
      rev::spark::SparkMax::PersistMode::kNoPersistParameters);

    rev::spark::SparkMaxConfig rightFollowConfig;
    rightFollowConfig.Follow(m_rightLeadMotor);

    m_rightFollowMotor.Configure(rightFollowConfig,
      rev::spark::SparkMax::ResetMode::kResetSafeParameters,
      rev::spark::SparkMax::PersistMode::kNoPersistParameters);
  }

  void TeleopPeriodic() {
    // Drive with arcade style
    m_robotDrive.ArcadeDrive(-m_stick.GetY(), m_stick.GetX());
  }
};

#ifndef RUNNING_FRC_TESTS
int main() { return frc::StartRobot<Robot>(); }
#endif
